core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -456,6 +457,58 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
   /**
+    * Callback that processes result of asynchronous version of conditionalUpdatePersistentPathAsync
+    */
+  private class conditionalUpdatePersistentPathCallback extends StatCallback {
+
+    def processResult(rc: Int, path: String, internalCtx: scala.Any, stat: Stat): Unit = {
+      val (zkUtils, zkHandle, path2, data, expectVersion, optionalCheckerAsync, callback, ctx) =
+        internalCtx.asInstanceOf[(ZkUtils, ZooKeeper, String, String, Int,
+          Option[(ZkUtils, String, String, (Boolean, Int, scala.Any) => Unit, scala.Any) => Unit],
+          (Boolean, Int, scala.Any) => Unit, scala.Any)]
+
+      Code.get(rc) match {
+        case Code.OK =>
+          debug("Conditional (async) update of path %s with value %s and expected version %d succeeded, returning the new version: %d"
+            .format(path, data, expectVersion, stat.getVersion))
+          return callback(true, stat.getVersion, ctx)
+        case Code.BADVERSION =>
+          optionalCheckerAsync match {
+            case Some(checker) =>
+              return checker(zkUtils, path, data, callback, ctx)
+            case _ => debug("Checker method is not passed skipping zkData match")
+          }
+          warn("Conditional update (async) of path %s with data %s and expected version %d failed due to bad version".format(path, data,
+            expectVersion))
+          return callback(false, -1, ctx)
+        case Code.CONNECTIONLOSS | Code.SESSIONEXPIRED =>
+          // retry
+          warn("Conditional update (async) of path %s with data %s and expected version %d failed due to connection/session loss %d. Retrying.".format(path, data,
+            expectVersion, Code.get(rc).intValue()))
+          zkHandle.setData(path, ZKStringSerializer.serialize(data), expectVersion, new conditionalUpdatePersistentPathCallback, internalCtx)
```

Comment:
@enothereska  I see, `zkClient` doesn't expose the async api directly, so you're bypassing the ZkClient wrapper and calling the zk handle directly, makes sense. If retrying is what ZkClient is doing, then keeping the same behavior sounds like the right thing to do. You may want to leave a comment explaining why you're dealing with the events this way.